storm-core/src/jvm/backtype/storm/blobstore/NimbusBlobStore.java

```diff
@@ -0,0 +1,419 @@
+import backtype.storm.generated.KeyAlreadyExistsException;
+import backtype.storm.generated.KeyNotFoundException;
+import backtype.storm.utils.NimbusClient;
+import backtype.storm.utils.Utils;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+public class NimbusBlobStore extends ClientBlobStore {
+  private static final Logger LOG = LoggerFactory.getLogger(NimbusBlobStore.class);
+
+  public class NimbusKeyIterator implements Iterator<String> {
+    private ListBlobsResult listBlobs = null;
+    private int offset = 0;
+    private boolean eof = false;
+    
+    public NimbusKeyIterator(ListBlobsResult listBlobs) {
+      this.listBlobs = listBlobs;
+      this.eof = (listBlobs.get_keys_size() == 0);
+    }
+    
+    private boolean isCacheEmpty() {
+      return listBlobs.get_keys_size() <= offset;
+    }
+
+    private void readMore() {
+      if (!eof) {
+        try {
+          offset = 0;
+          synchronized(client) {
+            listBlobs = client.getClient().listBlobs(listBlobs.get_session());
+          }
+          if (listBlobs.get_keys_size() == 0) {
+            eof = true;
+          }
+        } catch (TException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+    
+    @Override
+    public synchronized boolean hasNext() {
+      if (isCacheEmpty()) {
+        readMore();
+      }
+      return !eof;
+    }
+
+    @Override
+    public synchronized String next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      String ret = listBlobs.get_keys().get(offset);
+      offset++;
+      return ret;
+    }
+    
+    @Override
+    public void remove() {
+      throw new UnsupportedOperationException("Delete Not Supported");
+    }
+  }
+
+  public class NimbusDownloadInputStream extends InputStreamWithMeta {
+    private BeginDownloadResult beginBlobDownload;
+    private byte[] buffer = null;
+    private int offset = 0;
+    private int end = 0;
+    private boolean eof = false;
+
+    public NimbusDownloadInputStream(BeginDownloadResult beginBlobDownload) {
+      this.beginBlobDownload = beginBlobDownload;
+    }
+
+    @Override
+    public long getVersion() throws IOException {
+      return beginBlobDownload.get_version();
+    }
+
+    @Override
+    public synchronized int read() throws IOException {
+      if (isEmpty()) {
+        readMore();
+        if (eof) {
+          return -1;
+        }
+      }
+      int length = Math.min(1, available());
+      if (length == 0) {
+        return -1;
+      }
+      int ret = buffer[offset];
+      offset += length;
+      return ret;
+    }
+    
+    @Override 
+    public synchronized int read(byte[] b, int off, int len) throws IOException {
+      if (isEmpty()) {
+        readMore();
```

Comment:
Is it possible to just return the ByteBuffer to caller, to avoid the unnecessary double copying here?